x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java

```diff
@@ -1051,6 +1055,167 @@ public void testByteFieldWithIntegerSubfield() throws IOException {
+        } // find the first nested field starting from the end
+
+        StringBuilder stringPath = new StringBuilder(path.get(0).v1()); // the path we will ask for in the sql query
+        for (int i = 1; i <= endOfPathIndex + 2; i++) { // +2 because the index is now at the [index_of_a_nested_field]-1
+            if (path.get(i).v2() != NestedFieldType.LEAF || i == endOfPathIndex + 2) {
+                stringPath.append(".");
+                stringPath.append(path.get(i).v1());
+            }
+        }
+
+        Map<String, Object> expected = new HashMap<>();
+        expected.put("columns", singletonList(columnInfo("plain", stringPath.toString(), "keyword", JDBCType.VARCHAR, Integer.MAX_VALUE)));
+        expected.put("rows", singletonList(singletonList(randomValue.get())));
+        assertResponse(expected, runSql("SELECT " + stringPath.toString() + " FROM test"));
+    }
+
+    private enum NestedFieldType {
+        NESTED,
+        OBJECT,
+        LEAF;
+    }
+
+    private void addField(
+        XContentBuilder index,
+        boolean nestedFieldAdded,
+        int remainingFields,
+        List<Tuple<String, NestedFieldType>> path,
+        StringBuilder bulkContent,
+        Holder<String> randomValue
+    ) throws IOException {
+        String fieldName = randomAlphaOfLength(5);
+        String leafFieldName = randomAlphaOfLength(5);
+
+        // we need to make sure we add at least one nested field to the mapping, otherwise the test is not about nested fields
+        if (shouldAddNestedField() || (nestedFieldAdded == false && remainingFields == 1)) {
+            path.add(new Tuple<String, NestedFieldType>(fieldName, NestedFieldType.NESTED));
+            path.add(new Tuple<String, NestedFieldType>(leafFieldName, NestedFieldType.LEAF));
+            index.startObject(fieldName);
+            {
+                index.field("type", "nested");
+                index.startObject("properties");
+                {
+                    // A nested field always has a leaf field, even if not all nested fields in a path will have this value
+                    // indexed. We will index only the "leaf" field of the last nested field in the path, because this is the
+                    // one we will ask back from ES
+                    index.startObject(leafFieldName);
+                    {
+                        index.field("type", "keyword");
+                    }
+                    index.endObject();
+                    // from time to time set a null value instead of an actual value
+                    if (randomBoolean()) {
+                        randomValue.set(null);
+                        bulkContent.append("\"" + fieldName + "\":{\"" + leafFieldName + "\":null");
+                    } else {
+                        randomValue.set(randomAlphaOfLength(10));
+                        bulkContent.append("\"" + fieldName + "\":{\"" + leafFieldName + "\":\"" + randomValue.get() + "\"");
```

Comment:
I'm not sure I understand this entirely, but to me it seems the test will only add one value for the leaf field thats going to be requested and tested for here? Asking because its very common that there is more than one entry in a nested field, e.g. in a very simple case
```
     "fields": {
        "group" : ["fans"],
        "user": [{
            "first": ["John"],
            "last": ["Smith"],
          },
          {
            "first": ["Alice"],
            "last": ["White"],
          }
        ]
      }
```
where `user` is the nested field, and you want to retrieve `user.first`. I'm not sure this case is tested for, but I expect it to be common. Maybe less common but still relevant would be something like

```
     "fields": {
        "group" : ["fans"],
        "user": [{
            "first": ["John", "Jim", "James"],
            "last": ["Smith"],
          },
          {
            "first": ["Alice"],
            "last": ["White"],
          }
        ]
      }
```
where I would expect probably expect `user.first` to extract a list or array of four values. 
Maybe this is done in `AbstractFieldHitExtractor#unwrapFieldsMultiValue` but I think it would be great to have this explicitely covered by tests as well.